test/Test.hs

```diff
@@ -22,8 +44,413 @@ monadReader2 = assertEqual "should be equal" [(5, 0)] $
+        , testCase "msplit mzero :: strict StateT" $
+          let z :: SS.StateT Int [] String
+              z = mzero
+          in assertBool "strict StateT" $ null $ catMaybes $ SS.evalStateT (msplit z) 0
+        , testCase "msplit mzero :: lazy StateT" $
+          let z :: SL.StateT Int [] String
+              z = mzero
+          in assertBool "lazy StateT" $ null $ catMaybes $ SL.evalStateT (msplit z) 0
+        ]
+
+      , testGroup "msplit (return a `mplus` m) == return (Just a, m)" $
+        let sample = [1::Integer,2,3] in
+        [
+          testCase "msplit []" $ do
+            let op = sample
+                extract = fmap (fmap fst)
+            extract (msplit op) @?= [Just 1]
+            extract (msplit op >>= (\(Just (_,nxt)) -> msplit nxt)) @?= [Just 2]
+
+        , testCase "msplit ReaderT" $ do
+            let op = ask
+                extract = fmap fst . catMaybes . flip runReaderT sample
+            extract (msplit op) @?= [sample]
+            extract (msplit op >>= (\(Just (_,nxt)) -> msplit nxt)) @?= []
+
+        , testCase "msplit LogicT" $ do
+            let op :: LogicT [] Integer
+                op = foldr (mplus . return) mzero sample
+                extract = fmap fst . catMaybes . concat . observeAllT
+            extract (msplit op) @?= [1]
+            extract (msplit op >>= (\(Just (_,nxt)) -> msplit nxt)) @?= [2]
+
+        , testCase "msplit strict StateT" $ do
+            let op :: SS.StateT Integer [] Integer
+                op = (SS.modify (+1) >> SS.get `mplus` op)
+                extract = fmap fst . catMaybes . flip SS.evalStateT 0
+            extract (msplit op) @?= [1]
+            extract (msplit op >>= \(Just (_,nxt)) -> msplit nxt) @?= [2]
+
+        , testCase "msplit lazy StateT" $ do
+            let op :: SL.StateT Integer [] Integer
+                op = (SL.modify (+1) >> SL.get `mplus` op)
+                extract = fmap fst . catMaybes . flip SL.evalStateT 0
+            extract (msplit op) @?= [1]
+            extract (msplit op >>= \(Just (_,nxt)) -> msplit nxt) @?= [2]
+        ]
+      ]
+
+    , testGroup "fair disjunction"
+      [
+        -- base case
+        testCase "some odds"          $ [1,3,5,7] @=? observeMany 4 odds
+
+        -- without fairness, the second producer is never considered
+      , testCase "unfair disjunction" $ [1,3,5,7] @=? observeMany 4 oddsOrTwoUnfair
+
+        -- with fairness, the results are interleaved
+
+      , testCase "fair disjunction :: LogicT"   $ [1,2,3,5] @=? observeMany 4 oddsOrTwoFair
+
+        -- without fairness nothing would be produced, but with
+        -- fairness, a production is obtained
+
+      , testCase "fair production"   $ [2] @=? observeT oddsOrTwo
+
+        -- however, asking for additional productions will not
+        -- terminate (there are none, since the first clause generates
+        -- an infinity of mzero "failures")
+
+#if MIN_VERSION_base(4,9,0)
```

Comment:
```suggestion
#if MIN_VERSION_tasty_expected_failure(0,12,0)
```
Otherwise a future reader would be extremely confused what this test has to do with `base-4.9`.

Reply:
I agree that would be better but I don't think I can do that because tasty_expected_failure isn't compatible with those older versions so I had to remove it from the cabal file specification as well, which means cabal won't generate that definition.

Reply:
I see. You can still use `#ifdef MIN_VERSION_tasty_expected_failure`

Reply:
Nice call!  Done.